### ISSUE TITLE
SOA-EDIT-API should be allowed in auth metadata endpoint

### ISSUE THREAD

**Issue by toppk** [PowerDNS/pdns#8636]
- Program: Authoritative
 - Issue type: Bug report

### Short description

you can see SOA-EDIT-API in  /servers/localhost/zones/{zone}/metadata output but not in
                                                 /servers/localhost/zones/{zone}/metadata/SOA-EDIT-API

The documentation states that you cannot read or write this setting from the API.  I really don't understand why this setting is hidden from visibility (or for that matter blocked from writting).  At the very least it should be added to isValidMetadataKind or add a blacklist to the apiZoneMetadata GET routine

**Comment by Habbie** [comment]
It is exposed as the `soa_edit_api` property of a Zone object (https://doc.powerdns.com/authoritative/http-api/zone.html#zone). Leaving this open because we should perhaps improve the docs then, and perhaps fix the inconsistency between `isValidMetadataKind` and the GET output.

**Comment by zeha** [comment]
We also allow `SOA-EDIT` in the metadata API, probably by accident. That should get cleaned up too.

**Comment by james-stevens** [comment]
Personally, I think you have it backwards - instead of pretending metadata items aren't metadata, but are "zone properties", you should allow them to be addresses as the metadata items they really are (also as zone properties, if you insist).

Not being about to address some metadata items as metadata just creates a lot of exception code for both the client and server, which seems unnecessary.

IMHO

**Comment by kpfleming** [comment]
There's a related issue here: if SOA-EDIT-API is set to 'DEFAULT', it is not included in the 'zone properties' list when the zone is retrieved. When SOA-EDIT is set to 'DEFAULT' it *is* reported as a zone property. This causes the Ansible module that I wrote to always believe that it needs to set SOA-EDIT-API to 'DEFAULT' even though it already has that value.

**Comment by james-stevens** [comment]
> There's a related issue here: if SOA-EDIT-API is set to 'DEFAULT', it is not included in the 'zone properties' list when the zone is retrieved. When SOA-EDIT is set to 'DEFAULT' it _is_ reported as a zone property. This causes the Ansible module that I wrote to always believe that it needs to set SOA-EDIT-API to 'DEFAULT' even though it already has that value.

Right now some meta-data items can only be altered as zone properties and some can only be altered as meta-data items. Your code just has to be aware which is which and allow for it - for now - until all this is cleaned up.

https://github.com/james-stevens/powerdns-webui/blob/master/htdocs/index.html#L270

**Comment by kpfleming** [comment]
Right, I'm quite aware of that. The problem is that if SOA-EDIT-API has the value 'DEFAULT' it is not reported when the zone is retrieved, and if the code in question is trying to ensure that the zone's configuration matches a declared state (as is typical in Ansible), the code has to try to set it to 'DEFAULT' again. This doesn't cause any errors of course, and has no real effect other than wasting time by issuing an additional API call to set a property to the value it already has :-)

**Comment by james-stevens** [comment]
`SOA-EDIT-API` is unusual in that having no value & being set to `DEFAULT` have different meanings

https://doc.powerdns.com/authoritative/domainmetadata.html#soa-edit-api

So it should only report it has having no value, if it really has no value, *not* when its value is `DEFAULT`

**Comment by kpfleming** [comment]
Indeed, thanks for pointing this out, I found a bug in my Ansible module as a result :-)
